### DISCUSSION TITLE
Thread-safety of cel.Program

### DISCUSSION THREAD

**Discussion by yosiat** [google/cel-go#878] [Q&A] [answered]
Hi,

I have a cache for caching the evaluation of `Parse` so I don't think to parse the same expression over and over again.

I found that if I am caching `cel.Program` I get better performance - which makes sense to me.

My question - is it ok to use the same `cel.Program` and `Eval` on multiple go-routines on the same time?

I did fairly simple tests & parallel benchmarks with race detector running - and didn't got any race conditions - so looks like it's ok, but would like to get verification from the maintainers of the project.

**Comment by TristonianJones** [comment, marked as answer]
My apologies for not spotting this question sooner. The `cel.Program` is intended to be concurrency safe. There's no shared state within the `cel.Program`. If state tracking is enabled, then there's an allocation done per-evaluation to capture intermediate values from a single `Eval` call, but this state isn't shared either.

Hopefully that helps!
